File: computing_methods/CourseWork/src/Matrix.scala
class Matrix(h: Int, w: Int) {
  require(h > 0 && w > 0)
  private var values = Array.ofDim[Double](h, w)

  def this(a: Array[Array[Double]]) = {
    this(a.length, a(0).length)
    values = a.clone()
  }

  def getValues: Array[Array[Double]] = values.clone()

  def getHeight: Int = values.length

  def getWidth: Int = values(0).length

  def set(i: Int, j: Int, value: Double) = values(i) update(j, value)

  def apply(i: Int, j: Int) = get(i, j)

  def get(i: Int, j: Int): Double = values(i)(j)

  def norm = {
    var norm = 0d
    for (i <- 0 until getHeight) {
      var sum = 0d
      for (j <- 0 until getWidth) sum = sum + math.abs(get(i, j))
      if (sum > norm) norm = sum
    }
    norm
  }

  def mop(that: Matrix)(op: (Double, Double) => Double): Matrix = new Matrix((this.values zip that.values).map(t => (t._1 zip t._2).map(s => op(s._1, s._2))))

  def +(matrix: Matrix): Matrix = {
    assert(!(getHeight != matrix.getHeight || getWidth != matrix.getWidth))
    this.mop(matrix)(_ + _)
  }

  def *(matrix: Matrix): Matrix = {
    assert(getWidth == matrix.getHeight)
    val result = Array.ofDim[Double](getHeight, getWidth)
    for (i <- 0 until getHeight; j <- 0 until getWidth; k <- 0 until getWidth)
      result(i).update(j, result(i)(j) + get(i, k) * matrix(k, j))
    new Matrix(result)
  }

  def -(matrix: Matrix): Matrix = {
    assert(!(getHeight != matrix.getHeight || getWidth != matrix.getWidth))
    this.mop(matrix)(_ - _)
  }

  def *(v: Double): Matrix = {
    new Matrix(values.map(a => a.map(x => x * v)))
  }

  def *(v: Vector): Vector = {
    assert(getWidth == v.getDimension)
    val result = new Vector(getHeight)
    for (i <- 0 until getHeight; j <- 0 until getWidth) {
      result.set(i, result(i) + get(i, j) * v(j))
    }
    result
  }

  def croutLUDecomposition: (Matrix, Matrix) = {
    val n = getHeight
    val U = new Matrix(n, n)
    val L = new Matrix(n, n)
    for (i <- 0 until n) U.set(i, i, 1)
    for (j <- 0 until n) {
      for (i <- j until n) {
        var sum = 0d
        for (k <- 0 until j) sum = sum + L(i, k) * U(k, j)
        L.set(i, j, get(i, j) - sum)
      }
      for (i <- j until n) {
        var sum = 0d
        for (k <- 0 until j) sum = sum + L(j, k) * U(k, i)
        if (L(j, j) == 0) return null
        U.set(j, i, (get(j, i) - sum) / L(j, j))
      }
    }
    (U, L)
  }

  def determinant: Double = {
    val ul = croutLUDecomposition
    if (ul == null) return 0d
    var result = ul._1(0, 0) * ul._2(0, 0)
    for (i <- 1 until getHeight) result *= ul._1(i, i) * ul._2(i, i)
    result
  }

  def determinant1: Double = {
    if (getHeight == 1) get(0, 0)
    if (getHeight == 2) return get(0, 0) * get(1, 1) - get(1, 0) * get(0, 1)
    var result = 0d
    for (i <- 0 until getHeight) {
      result = result +
        (if (i % 2 == 0) 1 else  - 1) * get(i, 0) * minor(i, 0).determinant1
    }
    result
  }

  def reversed: Matrix = {
    val det: Double = determinant1
    //if (det.abs < 1e-18)
      //return null

    new Matrix(Array.tabulate(getHeight, getWidth)((j, i) => (if ((i + j) % 2 == 0) 1 else -1) * minor(i, j).determinant1 / det))
  }

  def minor(a: Int, b: Int): Matrix =
    new Matrix(Array.tabulate(getHeight - 1, getWidth - 1)((i, j) => get(if (i < a) i else i + 1, if (j < b) j else j + 1)))

  def conditionality = norm * reversed.norm

  def equals(matrix: Matrix, epsilon: Double): Boolean = {
    if (getHeight != matrix.getHeight || getWidth != matrix.getWidth) return false
    values.indices.forall((i) => values(i).indices.forall(j => (get(i, j) - matrix(i, j)).abs < epsilon))
  }

  def transpose: Matrix = new Matrix(values.transpose)

  def getDiagonal: Array[Double] = (for (i <- 0 until getHeight; x = get(i, i)) yield x).toArray

  def getRow(i: Int): Vector = new Vector(values(i).clone())

  def excludeVariableForward(i: Int): Matrix = {
    assert(!(i < 0 || i >= values(0).length))
    val result = getValues
    val row: Vector = getRow(i)
    for (k <- (i + 1 until values.length).reverse) {
      result update(k, (getRow(k) + (row * (-get(k, i) / get(i, i)))).getValues)
    }
    new Matrix(result)
  }

  def excludeVariableBackward(i: Int): Matrix = {
    assert(!(i < 0 || i >= values(0).length))
    val result = getValues
    var row: Vector = new Vector(getRow(i).getValues)
    row = row * (1d / row(i))
    for (j <- (0 until i).reverse) {
      result update(j, (getRow(j) + row * (-get(j, i))).getValues)
    }
    new Matrix(result)
  }

  def spectralRadiusLessOne(epsilon: Double): Double = {
    var result = 0d
    var nextResult = 0d
    val A = new Matrix(getValues)
    result = A.norm
    nextResult = result
    var Ak = A
    var i = 1
    val maxIterations = 10000
    do {
      result = nextResult
      Ak = Ak * A
      i = i + 1
      nextResult = math.pow(Ak.norm, 1d / i)
    } while (math.abs(result - nextResult) > epsilon && i < maxIterations)
    nextResult
  }

  override def toString = {
    val result: StringBuilder = new StringBuilder
    for (i <- 0 until getHeight) {
      for (j <- 0 until getWidth)
        result.append("%3.3f  " format get(i, j))
      result.append("\n")
    }
    result.toString()
  }
}

object Matrix {
  def zero(h: Int, w: Int): Matrix = new Matrix(h, w)

  def zero(n: Int): Matrix = zero(n, n)

  def byDiagonal(trace: Array[Double]): Matrix = new Matrix(Array.tabulate[Double](trace.length, trace.length)((i, j) => if (i == j) trace(i) else 0d))

  def unit(d: Int): Matrix = new Matrix(Array.tabulate[Double](d, d)((i, j) => if (i == j) 1d else 0d))

  def getExtended(a: Matrix, b: Vector): Matrix = new Matrix(Array.tabulate[Double](a.getHeight, a.getWidth + 1)((i, j) => if (j < a.getWidth) a(i, j) else b(i)))

  def parse(s: String): Matrix = {
    var split = scala.Vector[Int]()
    var balance = 0
    for (i <- 0 until s.length)
      s(i) match {
        case '{' => balance = balance + 1
        case '}' => balance = balance - 1
        case ',' => if (balance == 0) split = split :+ i
        case _ =>
      }
    split = split :+ s.length
    new Matrix((for {i <- split.indices
                     line = s.substring((if (i == 0) 0 else split(i - 1)) + 1, split(i) - 1).trim.replaceAll("[\\}\\{,]", " ").split("\\s+")
                     r = Array.tabulate[Double](line.length - 1)(j => line(j + 1).toDouble)
    } yield r).toArray)
  }
}

File: computing_methods/SystemOfLinearEquations/src/Matrix.java
import javafx.util.Pair;

import java.util.ArrayList;
import java.util.Arrays;

public class Matrix {
    private double[][] values;

    public Matrix(int height, int width) {
        assert !(height <= 0 || width <= 0);
        this.values = new double[height][width];
    }

    public Matrix(double[]... values) {
        assert !(values == null || values.length == 0);
        int width = values[0].length;
        for (double[] component : values)
            assert component.length == width;
        this.values = values;
    }

    public double[][] getValues() {
        double[][] result = new double[values.length][];
        for (int i = 0; i < result.length; i++) result[i] = Arrays.copyOf(values[i], values[i].length);
        return result;
    }

    public int getHeight() {
        return values.length;
    }

    public int getWidth() {
        return values[0].length;
    }

    public void swapRows(int i, int j) {
        if (i == j) return;
        for (int k = 0; k < getWidth(); k++) {
            double temp = values[i][k];
            values[i][k] = values[j][k];
            values[j][k] = temp;
        }
    }

    public void set(int i, int j, double value) {
        values[i][j] = value;
    }

    public double get(int i, int j) {
        return values[i][j];
    }

    public boolean equals(Matrix matrix, double epsilon) {
        if (getHeight() != matrix.getHeight() || getWidth() != matrix.getWidth()) return false;
        for (int i = 0; i < getHeight(); i++)
            for (int j = 0; j < getWidth(); j++)
                if (!(Math.abs(get(i, j) - matrix.get(i, j)) < epsilon))
                    return false;
        return true;
    }

    public double norm() {
        double norm = 0.;
        for (int i = 0; i < getHeight(); i++) {
            double sum = 0.;
            for (int j = 0; j < getWidth(); j++) sum += Math.abs(get(i, j));
            if (sum > norm) norm = sum;
        }
        return norm;
    }

    public double conditionality() {
        return norm() * reversed().norm();
    }

    public Matrix add(Matrix matrix) {
        assert !(getHeight() != matrix.getHeight() || getWidth() != matrix.getWidth());
        double[][] result = new double[getHeight()][getWidth()];
        for (int i = 0; i < getHeight(); i++) {
            for (int j = 0; j < result[i].length; j++) result[i][j] = get(i, j) + matrix.get(i, j);
        }
        return new Matrix(result);
    }

    public Matrix subtract(Matrix b) {
        return add(b.multiply(-1.));
    }

    public Matrix multiply(Matrix matrix) {
        assert getWidth() == matrix.getHeight();
        double[][] result = new double[getHeight()][matrix.getWidth()];
        for (int i = 0; i < getHeight(); i++) {
            for (int j = 0; j < matrix.getWidth(); j++)
                for (int k = 0; k < getWidth(); k++) result[i][j] += get(i, k) * matrix.get(k, j);
        }
        return new Matrix(result);
    }

    public Matrix multiply(double v) {
        double[][] result = new double[getHeight()][getWidth()];
        for (int i = 0; i < getHeight(); i++) {
            for (int j = 0; j < getWidth(); j++) result[i][j] = get(i, j) * v;
        }
        return new Matrix(result);
    }

    public Vector multiply(Vector vector) {
        assert getWidth() == vector.getDimensions();
        double[] result = new double[getHeight()];
        for (int i = 0; i < getHeight(); i++)
            for (int j = 0; j < getWidth(); j++) result[i] += get(i, j) * vector.get(j);
        return new Vector(result);
    }

    /*public double determinant() {
        assert getHeight() == getWidth();
        if (getHeight() == 1) return get(0, 0);
        if (getHeight() == 2) return get(0, 0) * get(1, 1) - get(1, 0) * get(0, 1);
        double result = 0.;
        for (int i = 0; i < getHeight(); i++) {
            double temp = get(i, 0) * minor(i, 0).determinant();
            result += (i % 2 == 0 ? 1 : -1) * temp;
        }
        return result;
    }*/

    public Pair<Matrix, Matrix> croutLUDecomposition() {
        int n = getHeight();
        Matrix U = new Matrix(n, n);
        Matrix L = new Matrix(n, n);
        double sum;
        for (int i = 0; i < n; i++) U.set(i, i, 1);
        for (int j = 0; j < n; j++) {
            for (int i = j; i < n; i++) {
                sum = 0;
                for (int k = 0; k < j; k++) sum = sum + L.get(i, k) * U.get(k, j);
                L.set(i, j, get(i, j) - sum);
            }
            for (int i = j; i < n; i++) {
                sum = 0;
                for (int k = 0; k < j; k++) sum = sum + L.get(j, k) * U.get(k, i);
                if (L.get(j, j) == 0) return null;
                U.set(j, i, (get(j, i) - sum) / L.get(j, j));
            }
        }
        return new Pair<>(U, L);
    }

    public double determinant() {
        double result;
        Pair<Matrix, Matrix> UL = croutLUDecomposition();
        if (UL == null) return 0.;
        result = UL.getKey().get(0, 0) * UL.getValue().get(0, 0);
        for (int i = 1; i < getHeight(); i++)
            result *= UL.getKey().get(i, i) * UL.getValue().get(i, i);
        return result;
    }

    public Matrix reversed() {
        double determinant = determinant();
        if (Math.abs(determinant) < 1e-18)
            return null;
        double[][] result = new double[getWidth()][getHeight()];
        for (int i = 0; i < getHeight(); i++) {
            for (int j = 0; j < getWidth(); j++) {
                double temp = minor(i, j).determinant() / determinant;
                result[j][i] = ((i + j) % 2 == 0 ? 1 : -1) * temp;
            }
        }
        return new Matrix(result);
    }

    public Matrix transpose() {
        double[][] result = new double[getWidth()][getHeight()];
        for (int i = 0; i < getWidth(); i++) {
            for (int j = 0; j < getHeight(); j++) result[i][j] = get(j, i);
        }
        return new Matrix(result);
    }

    public Matrix minor(int a, int b) {
        double[][] result = new double[getHeight() - 1][getWidth() - 1];
        for (int i = 0; i < result.length; i++)
            for (int j = 0; j < result[i].length; j++) result[i][j] = get(i < a ? i : i + 1, j < b ? j : j + 1);
        return new Matrix(result);
    }

    public static Matrix unit(int dimensions) {
        double[][] result = new double[dimensions][dimensions];
        for (int i = 0; i < dimensions; i++)
            for (int j = 0; j < dimensions; j++)
                result[i][j] = i == j ? 1. : 0.;
        return new Matrix(result);
    }

    public static Matrix zero(int dimensions) {
        return zero(dimensions, dimensions);
    }

    public static Matrix zero(int height, int width) {
        return new Matrix(new double[height][width]);
    }

    public double[] getDiagonal() {
        double[] result = new double[getHeight()];
        for (int i = 0; i < getHeight(); i++) result[i] = get(i, i);
        return result;
    }

    public static Matrix byDiagonal(double[] trace) {
        double[][] result = new double[trace.length][trace.length];
        for (int i = 0; i < result.length; i++) result[i][i] = trace[i];
        return new Matrix(result);
    }

    public Vector getRow(int i) {
        assert !(i < 0 || i >= values.length);
        return new Vector(values[i].clone());
    }

    public Matrix excludeVariableForward(int i) {
        assert !(i < 0 || i >= values[0].length);
        double[][] result = getValues();
        Vector row = getRow(i);
        for (int k = values.length - 1; k >= i + 1; k--)
            result[k] = getRow(k).add(row.multiply(-get(k, i) / get(i, i))).getValues();
        return new Matrix(result);
    }

    public Matrix excludeVariableBackward(int i) {
        assert !(i < 0 || i >= values[0].length);
        double[][] result = getValues();
        Vector row = new Vector(getRow(i).getValues());
        row = row.multiply(1 / row.get(i));
        for (int j = i - 1; j >= 0; --j) result[j] = getRow(j).add(row.multiply(-get(j, i))).getValues();
        return new Matrix(result);
    }

    public static Matrix getExtended(Matrix a, Vector b) {
        double[][] result = new double[a.getHeight()][a.getWidth() + 1];
        for (int i = 0; i < a.getHeight(); i++) {
            for (int j = 0; j < a.getWidth(); j++)
                result[i][j] = a.get(i, j);
            result[i][result[i].length - 1] = b.get(i);
        }
        return new Matrix(result);
    }


    public static Matrix parse(String s) {
        ArrayList<Integer> splitPoints = new ArrayList<>();
        int bracesBalance = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '{') {
                ++bracesBalance;
            } else if (s.charAt(i) == '}') {
                --bracesBalance;
            } else if (s.charAt(i) == ',' && bracesBalance == 0) {
                splitPoints.add(i);
            }
        }
        splitPoints.add(s.length());
        double[][] result = new double[splitPoints.size()][];
        for (int i = 0; i < splitPoints.size(); i++) {
            String[] line = s.substring(i == 0 ? 0 : splitPoints.get(i - 1) + 1, splitPoints.get(i) - 1).trim()
                    .replaceAll("[\\}\\{,]", " ").split("\\s+");
            result[i] = new double[line.length - 1];
            for (int j = 1; j < line.length; j++) {
                result[i][j - 1] = Double.parseDouble(line[j]);
            }
        }
        return new Matrix(result);
    }

    public static final int TO_STRING_DIGITS = 3;

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < getHeight(); i++) {
            for (int j = 0; j < getWidth(); j++)
                result.append(String.format("% " + TO_STRING_DIGITS + "." + TO_STRING_DIGITS + "f ", get(i, j)));
            result.append("\n");
        }
        return result.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (!(o instanceof Matrix)) return false;
        Matrix matrix = (Matrix) o;
        if (getHeight() != matrix.getHeight() || getWidth() != matrix.getWidth()) return false;
        for (int i = 0; i < getHeight(); i++)
            for (int j = 0; j < getWidth(); j++) if (get(i, j) != matrix.get(i, j)) return false;
        return true;
    }
}
